### PR TITLE
Change TAG handling

### PR DESCRIPTION
**Pull request by derletztekick**
Storage of tags are moved in seperate table as suggested in https://github.com/ilosuna/mylittleforum/issues/289 and https://github.com/ilosuna/mylittleforum/pull/287

### CODE REVIEW

**Comment by derletztekick** [comment]
@auge8472 Please check my changes and report problems

**Comment by auge8472** [comment]
I will review the code at the weekend. Christmas time is pure stress.

**Comment by derletztekick** [comment]
> Christmas time is pure stress.

Einen schönen ersten Advent. ;-)

**Comment by auge8472** [comment]
> Einen schönen ersten Advent. ;-)

Der Wunsch geht zurück. Wie ich feststellen darf, ist die Kapselung ein nicht unerheblicher Aufwand. Ich habe bisher gerade einmal 2 Dateien geschafft.

**Comment by derletztekick** [comment]
> Wie ich feststellen darf, ist die Kapselung ein nicht unerheblicher Aufwand

Okay, but that was predictable, wasn't it? The handling of both, tags _and_ postings, was changed.

File: includes/posting.inc.php [hunk 1]
```diff
@@ -1,1365 +1,1646 @@
...
-	require('modules/captcha/captcha.php');
-	$captcha = new Captcha();
-}
-
-// reorder categories if user has a personal category selection (selected categories first):
-if (isset($category_selection)) {
-	// put selected cegories in $reordered_categories:
-	foreach ($category_selection as $cat) {
-		$reordered_categories[$cat] = $categories[$cat];
+	if (!defined('IN_INDEX')) {
```
Comment (auge8472) [inline includes/posting.inc.php]: Warum die erhöhte Einrückung? Ist da etwas schiefgelaufen?

File: includes/bookmark.inc.php [hunk 2]
```diff
@@ -160,31 +158,25 @@
...
-				if (my_strlen($tags, $lang['charset']) > 253) 
-					$errors[] = 'error_bookmark_tags_too_long';
-				
-				// Taken from posting.inc.php
-				$tags       = str_replace(';', '', $tags);
-				$tags       = str_replace('"', '', $tags);
-				$tags_array = explode(',', $tags);
-				$tags       = ';';
-				foreach ($tags_array as $tag) {
+				$tagStr = trim($_POST['tags']);
```
[the remark this answers is not in the document]
Comment (derletztekick) [reply includes/bookmark.inc.php]: You review an outdated file! I remove the variable $tagStr.

File: includes/posting.inc.php (hunk 1 above)
Comment (derletztekick) [reply includes/posting.inc.php]: I used a online service to re-format the file.

File: install/install.sql [hunk 3]
```diff
@@ -11,9 +11,11 @@ CREATE TABLE mlf2_useronline (ip char(15) NOT NULL default '',time int(14) NOT N
...
 CREATE TABLE mlf2_entries_cache (cache_id int(11) NOT NULL, cache_text mediumtext NOT NULL, PRIMARY KEY (cache_id)) CHARSET=utf8 COLLATE=utf8_general_ci;
 CREATE TABLE mlf2_userdata_cache (cache_id int(11) NOT NULL, cache_signature text NOT NULL, cache_profile text NOT NULL, PRIMARY KEY (cache_id)) CHARSET=utf8 COLLATE=utf8_general_ci;
-CREATE TABLE mlf2_bookmarks (id int(11) NOT NULL AUTO_INCREMENT, user_id int(11) NOT NULL, posting_id int(11) NOT NULL, time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP, subject varchar(255) NOT NULL, tags varchar(255) NOT NULL default '', order_id int(11) NOT NULL DEFAULT '0', PRIMARY KEY (id), UNIQUE KEY UNIQUE_uid_pid (user_id,posting_id)) ENGINE=InnoDB CHARSET=utf8 COLLATE=utf8_general_ci;
+CREATE TABLE mlf2_bookmarks (id int(11) NOT NULL AUTO_INCREMENT, user_id int(11) NOT NULL, posting_id int(11) NOT NULL, time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP, subject varchar(255) NOT NULL, order_id int(11) NOT NULL DEFAULT '0', PRIMARY KEY (id), UNIQUE KEY UNIQUE_uid_pid (user_id,posting_id)) ENGINE=InnoDB CHARSET=utf8 COLLATE=utf8_general_ci;
```
[the remark this answers is not in the document]
Comment (derletztekick) [reply install/install.sql]: USING BTREE was removed

File: install/install.sql [hunk 4]
```diff
@@ -11,9 +11,11 @@ CREATE TABLE mlf2_useronline (ip char(15) NOT NULL default '',time int(14) NOT N
...
 CREATE TABLE mlf2_read_entries (user_id int(11) UNSIGNED NOT NULL, posting_id int(11) UNSIGNED NOT NULL, time timestamp NOT NULL, PRIMARY KEY (user_id, posting_id), KEY `user_id` (`user_id`), KEY `posting_id` (`posting_id`)) ENGINE=InnoDB CHARSET=utf8 COLLATE=utf8_general_ci;
 CREATE TABLE mlf2_temp_infos (name varchar(50) NOT NULL, value varchar(255) NOT NULL, time timestamp NULL DEFAULT NULL, PRIMARY KEY (name)) ENGINE=InnoDB CHARSET=utf8 COLLATE=utf8_general_ci;
+CREATE TABLE mlf2_bookmark_tags (`bid` int(11) NOT NULL, `tid` int(11) NOT NULL), PRIMARY KEY (`bid`,`tid`) USING BTREE CHARSET=utf8 COLLATE=utf8_general_ci;
+CREATE TABLE mlf2_tags (`id` int(11) NOT NULL, `tag` varchar(255) NOT NULL), PRIMARY KEY (`id`), UNIQUE KEY `tag` (`tag`) CHARSET=utf8 COLLATE=utf8_general_ci;
```
[the remark this answers is not in the document]
Comment (derletztekick) [reply install/install.sql]: You review an outdated file!
USING BTREE was removed.

File: update/update.sql [hunk 5]
```diff
@@ -358,6 +358,8 @@ DELETE FROM mlf2_settings WHERE name = 'read_state_expiration_date'
 --------------------------------------------------------------------------------
 -- 2.4.5 to 2.4.6
 /*
-ALTER TABLE mlf2_bookmarks ADD tags VARCHAR(255) CHARACTER SET utf8 COLLATE utf8_general_ci NOT NULL DEFAULT '';
+CREATE TABLE mlf2_bookmark_tags (`bid` int(11) NOT NULL, `tid` int(11) NOT NULL), PRIMARY KEY (`bid`,`tid`) USING BTREE CHARSET=utf8 COLLATE=utf8_general_ci;
+CREATE TABLE mlf2_tags (`id` int(11) NOT NULL, `tag` varchar(255) NOT NULL), PRIMARY KEY (`id`), UNIQUE KEY `tag` (`tag`) CHARSET=utf8 COLLATE=utf8_general_ci;
```
[the remark this answers is not in the document]
Comment (derletztekick) [reply update/update.sql]: You review an outdated file!
USING BTREE was removed.

File: update/update_2.3.5-2.4.php [hunk 6]
```diff
@@ -384,8 +384,65 @@
...
+			fwrite($db_settings_file, "\$db_settings['bookmark_tags_table']  = '". addslashes($db_settings['bookmark_tags_table']) ."';\n");
+			fwrite($db_settings_file, "\$db_settings['tags_table']           = '". addslashes($db_settings['tags_table']) ."';\n");
+			
+			fwrite($db_settings_file, "\$db_settings['read_status_table']    = '". addslashes($db_settings['read_status_table']) ."';\n");
+			fwrite($db_settings_file, "\$db_settings['temp_infos_table']     = '". addslashes($db_settings['temp_infos_table']) ."';\n");
+			fwrite($db_settings_file, "?".">\n");
+			flock($db_settings_file, 3);
+			fclose($db_settings_file);
+			
+			// new tables
```
[the remark this answers is not in the document]
Comment (derletztekick) [reply update/update_2.3.5-2.4.php]: USING BTREE was removed.

[the remark this answers is not in the document]
Comment (auge8472) [reply update/update_2.3.5-2.4.php]: Please add ` ENGINE=InnoDB` for the new tables as the last step. Otherwise I see no need for further changes.

**Comment by derletztekick** [comment]
I check the install and update routine and found some mistakes. :-/ I hope, I could handle every issue.

**Pull request merged**